File: runtime/python/units/wasi/fd_write.py
# requires: memory/iwl, memory/iws, memory/read_string
def wasi_fd_write(self, fd, iovs_ptr, iovs_len, nwritten_ptr):
    io = self.fds.get(fd)
    if io is None or isinstance(io, self.WasiDir):
        return self.ERRNO_BADF
    if not (self.fd_meta[fd][0] & self.RIGHTS_FD_WRITE):
        return self.ERRNO_NOTCAPABLE
    written = 0
    try:
        # fdflags::APPEND is honoured here (not via O_APPEND on the OS handle) so fd_fdstat_set_flags(0) can turn it back off.
        if self.fd_meta[fd][2] & 0x1 and io not in self.std_ios:
            io.seek(0, os.SEEK_END)
        for i in range(iovs_len):
            ptr = self.memory.iwl(iovs_ptr + i * 8)
            length = self.memory.iwl(iovs_ptr + i * 8 + 4)
            written += io.write(self.memory.read_string(ptr, length))
        io.flush()
    except OSError:
        return self.ERRNO_IO
    self.memory.iws(nwritten_ptr, written)
    return self.ERRNO_SUCCESS

File: runtime/go/units/wasi/fd_write.go
// requires: memory/i32_load, memory/i32_store, memory/read_string
func (w *WASI) wasi_fd_write(fd, iovsPtr, iovsLen, nwrittenPtr uint32) uint32 {
    f, ok := w.fds[fd].(*os.File)
    if !ok {
        return wasiBadf
    }
    if e := w.checkRight(fd, rightFdWrite); e != wasiOk {
        return e
    }
    // APPEND is honored here rather than by opening the OS fd O_APPEND, so
    // fd_fdstat_set_flags can turn it off at runtime: seek to end
    // before writing.
    if m, ok := w.meta[fd]; ok && m.fdflags&fdflagAppend != 0 && !w.isStdio(f) {
        f.Seek(0, 2) // whence 2 = io.SeekEnd
    }
    written := uint32(0)
    for i := uint32(0); i < iovsLen; i++ {
        ptr := w.memory.i32_load(uint64(iovsPtr) + uint64(i)*8)
        length := w.memory.i32_load(uint64(iovsPtr)+uint64(i)*8 + 4)
        n, err := f.Write(w.memory.read_string(uint64(ptr), uint64(length)))
        written += uint32(n)
        if err != nil {
            return wasiIo
        }
    }
    w.memory.i32_store(uint64(nwrittenPtr), written)
    return wasiOk
}
